pkg/controller/nodeipam/ipam/multicidrset/multi_cidr_set.go

```diff
@@ -0,0 +1,365 @@
+		if cidr.IP.To4() != nil {
+			ipInt := binary.BigEndian.Uint32(cidr.IP) | (^binary.BigEndian.Uint32(cidr.Mask))
+			binary.BigEndian.PutUint32(ip, ipInt)
+		} else {
+			// ipIntLeft          |         ipIntRight
+			// 2001:0DB8:1234:0000:0000:0000:0000:0000
+			ipIntLeft := binary.BigEndian.Uint64(cidr.IP[:net.IPv6len/2]) | (^binary.BigEndian.Uint64(cidr.Mask[:net.IPv6len/2]))
+			ipIntRight := binary.BigEndian.Uint64(cidr.IP[net.IPv6len/2:]) | (^binary.BigEndian.Uint64(cidr.Mask[net.IPv6len/2:]))
+			binary.BigEndian.PutUint64(ip[:net.IPv6len/2], ipIntLeft)
+			binary.BigEndian.PutUint64(ip[net.IPv6len/2:], ipIntRight)
+		}
+		end, err = s.getIndexForCIDR(&net.IPNet{
+			IP:   net.IP(ip).Mask(s.nodeMask),
+			Mask: s.nodeMask,
+		})
+		if err != nil {
+			return -1, -1, err
+		}
+	}
+	return begin, end, nil
+}
+
+// Release releases the given CIDR range.
+func (s *MultiCIDRSet) Release(cidr *net.IPNet) error {
+	begin, end, err := s.getBeginningAndEndIndices(cidr)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	for i := begin; i <= end; i++ {
+		// Remove from the allocated CIDR Map and decrement the counter only if currently
+		// marked allocated. Avoids double counting.
+		currCIDR, err := s.indexToCIDRBlock(i)
+		if err != nil {
+			return err
+		}
+		if _, ok := s.AllocatedCIDRMap[currCIDR.String()]; ok {
+			delete(s.AllocatedCIDRMap, currCIDR.String())
+			s.allocatedCIDRs--
+			cidrSetReleases.WithLabelValues(s.Label).Inc()
+		}
+	}
+
+	cidrSetUsage.WithLabelValues(s.Label).Set(float64(s.allocatedCIDRs) / float64(s.MaxCIDRs))
+
+	return nil
+}
+
+// Occupy marks the given CIDR range as used. Occupy succeeds even if the CIDR
+// range was previously used.
+func (s *MultiCIDRSet) Occupy(cidr *net.IPNet) (err error) {
+	begin, end, err := s.getBeginningAndEndIndices(cidr)
+	if err != nil {
+		return err
+	}
+	s.Lock()
+	defer s.Unlock()
+
+	for i := begin; i <= end; i++ {
+		// Add to the allocated CIDR Map and increment the counter only if not already
+		// marked allocated. Avoids double counting.
+		currCIDR, err := s.indexToCIDRBlock(i)
+		if err != nil {
+			return err
+		}
+		if _, ok := s.AllocatedCIDRMap[currCIDR.String()]; !ok {
+			s.AllocatedCIDRMap[currCIDR.String()] = true
+			cidrSetAllocations.WithLabelValues(s.Label).Inc()
+			s.allocatedCIDRs++
+		}
+	}
+	cidrSetUsage.WithLabelValues(s.Label).Set(float64(s.allocatedCIDRs) / float64(s.MaxCIDRs))
+
+	return nil
+}
+
+func (s *MultiCIDRSet) getIndexForCIDR(cidr *net.IPNet) (int, error) {
+	return s.getIndexForIP(cidr.IP)
+}
+
+func (s *MultiCIDRSet) getIndexForIP(ip net.IP) (int, error) {
+	if ip.To4() != nil {
+		cidrIndex := (binary.BigEndian.Uint32(s.ClusterCIDR.IP) ^ binary.BigEndian.Uint32(ip.To4())) >> uint32(32-s.NodeMaskSize)
+		if cidrIndex >= uint32(s.MaxCIDRs) {
+			return 0, fmt.Errorf("CIDR: %v/%v is out of the range of CIDR allocator", ip, s.NodeMaskSize)
+		}
+		return int(cidrIndex), nil
+	}
+	if ip.To16() != nil {
```

Comment:
an IPv4 address returns != nil in To16()

https://go.dev/play/p/8Lo4lSmegex

we have some helpers in k8s.io/utils to help with this, netutils.IsIPv4() and IsIPv6()

Reply:
nevermind, this was this way in the old cidrset, the the ip.To4() returns

Reply:
Borrowed this code from the existing rangeallocator: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/ipam/cidrset/cidr_set.go#L101

We need to update that as well, may be in a separate PR though